src/persist-client/src/operators/shard_source.rs

```diff
@@ -546,30 +550,144 @@ where
                 .await
         };
 
-        while let Some(event) = descs_input.next_mut().await {
-            if let Event::Data(cap, data) = event {
-                // `LeasedBatchPart`es cannot be dropped at this point w/o
-                // panicking, so swap them to an owned version.
-                for (_idx, part) in data.drain(..) {
-                    let (leased_part, fetched) = fetcher
-                        .fetch_leased_part(fetcher.leased_part_from_exchangeable(part))
-                        .await;
-                    let fetched = fetched.expect("shard_id should match across all workers");
+        // We use two buffers to queue up and pipeline reads to Blob, to ensure we can adhere
```

Comment:
I haven't looked at the details here yet, but I have a pretty strong preference that if we're going to introduce pipelined fetches, that we do it by pulling out a new abstraction instead of inlining the complexity into the operator impl. IMO these operators are already over the limit of reasonable cognitive complexity (some inherent, some perhaps not), so I'd be regretful to increase that any more, if we can avoid it

in my head that abstraction looks something almost like a channel in that there is one handle where we push in new things to fetch (probably a non-async fn) and another handle that pulls out things that have finished fetching (probably an async fn that returns `Option<FetchedThing>`). all the various tunables could then be passed in as constructor params

however, I'm less opinionated on the specifics of the abstraction than I am about there being one. this also makes it mush easier to write unit tests for

Reply:
👍 have this as a WIP, a lot of the logic here can be pulled out into a separate abstraction

Reply:
@danhhz pulled this into a new bit `PipelineFetcher`. Your idea of the API is virtually the same as the loop I had before with a non-blocking input and async output handle. It's now abstracted out, though still pretty attached to what `shard_source` needs, curious to get your thoughts on it and if it hits the right level of abstraction